apps/enphasesolar/enphase_solar.star

```diff
@@ -0,0 +1,174 @@
+AUTH_URL = "https://api.enphaseenergy.com/oauth/token"
+EXPIRE_MSG = "Access token has expired"
+
+# Due to limited number of API calls in free version of Enphase Application, here it only update
+# information every hour.
+TTL_SECONDS = 3600
+
+def check_response(response):
+    if response.status_code == 200:
+        return 200, response.json()
+    elif response.status_code == 401 and EXPIRE_MSG in response.json().get("details"):
+        print("Access token is expired and refresh token now. Call next time.")
+        return 401, None
+    else:
+        print("Request is failed with status code {} {}".format(response.status_code, response.json()))
+        return response.status_code, None
+
+def get_system_stats(api_key):
+    headers = {
+        "Authorization": "Bearer " + cache.get("access_token"),
+        "Content-Type": "application/json",
+    }
+
+    params = {
+        "key": api_key,
+    }
+
+    response = http.get(ENDPOINT, params = params, headers = headers)
+
+    status, result = check_response(response)
+
+    if status == 200 and result:
+        return 200, str(result["systems"][0]["energy_today"])
+    elif status == 401:
+        return 401, None
+    else:
+        msg = "No result returned."
+        return status, msg
+
+def request_refresh_token(refresh_token_code, client_id, client_secret):
+    encoded_client_secrets = base64.encode("{}:{}".format(client_id, client_secret))
+    headers = {
+        "Authorization": "Basic {}".format(encoded_client_secrets),
+        "Content-Type": "application/json",
+    }
+    params = {
+        "grant_type": "refresh_token",
+        "refresh_token": refresh_token_code,
+    }
+
+    response = http.post(AUTH_URL, params = params, headers = headers)
+    if response.status_code == 200:
+        print("Refresh token successfully.")
+        cache.set("access_token", response.json()["access_token"], ttl_seconds = int(response.json()["expires_in"]))
+        cache.set("refresh_token", response.json()["refresh_token"])
+    else:
+        msg = "Refresh token failed with status code {}, message {}".format(
+            response.status_code,
+            response.json(),
+        )
+        render_msg(msg)
+
+def render_msg(msg):
+    """Render message to App"""
+    return render.Root(
+        child = render.WrappedText(
+            content = msg,
+            width = 50,
+            color = "#fa0",
+        ),
+    )
+
+def main(config):
+    """Display Energy produced by solar panels daily basis."""
+    if not config.bool("switch", True):
+        return render_msg("Switch is off, so take a rest.")
+
+    access_token = config.str("access_token")
+    refresh_token = config.str("refresh_token")
+    client_id = config.str("client_id")
+    client_secret = config.str("client_secret")
+    api_key = config.str("api_key")
+
+    if not all([access_token, refresh_token, client_id, client_secret, api_key]):
+        return render_msg("Missing credential information.")
```

Comment:
It'd be great if you could provide a default graphic for the case when no configuration is provided. That would allow users to see a preview of the app, before they decided to go looking for all the require config to set it all up.

Reply:
I am not sure how to provide a default graphic, but I add more information in the msg rendered including what information this app provides.

Reply:
The way most apps do it is to check if the config params are present, and if they're not just use some mock data to render what the app would look like if it had config. Could be good to add that in a later PR, if you're up for it. =)

Reply:
Aha, gotcha. I will address this in next PR! We are getting over 50 KWH from the solar panels now, still cold in Boston. Super excited to get the this app running on Tidbyt device without running my code from my laptop and push to the device using a script.